File: incre-tests/synduce/list/mts_and_mps_hom.f
Config ExtraGrammar = "AutoLifter";
Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | single Int | concat {CList, CList};

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: List.
let res = fix (
  \f: List -> {Int, Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0, 0}
  | cons {hd, tl} ->
    let result = f tl in
    let new_sum = + hd result.1 in
    let new_mps = max 0 (+ hd result.2) in
    let new_mts = max result.3 (+ hd result.1) in
    {new_sum, new_mps, new_mts}
  end
) xs in {res.2, res.3};

cat = fix (
  \f: List -> List -> List. \xs: List. \ys: List.
  match xs with
    nil _ -> ys
  | cons {a, b} -> cons {a, f b ys}
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    cnil _ -> nil unit
  | single a -> cons {a, nil unit}
  | concat {a, b} -> cat (f a) (f b)
  end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/mts_and_mps_hom.ml
config EnableDeepCoder = true

type list = Nil of unit | Cons of int * list
type clist = CNil of unit | Single of int | Concat of clist * clist

let max a b = if a < b then b else a

let spec xs =
  let rec f xs =
    match xs with
    | Nil _ -> (0, 0, 0)
    | Cons (hd, tl) ->
      let result = f tl in
      match result with
      | (sum, mps, mts) ->
        let new_sum = hd + sum in
        let new_mps = max 0 (hd + mps) in
        let new_mts = max mts (hd + sum) in
        (new_sum, new_mps, new_mts)
  in
  let res = f xs in
  match res with
  | (_, mps, mts) -> (mps, mts)

let rec cat xs ys =
  match xs with
  | Nil _ -> ys
  | Cons (a, b) -> Cons (a, cat b ys)

val repr: clist -> list compress
let rec repr xs =
  match xs with
  | CNil _ -> Nil ()
  | Single a -> Cons (a, Nil ())
  | Concat (a, b) -> cat (repr a) (repr b)

let program x = spec (repr x)